cloudmarker/stores/mongodbstore.py

```diff
@@ -0,0 +1,100 @@
+            db (str): MongoDB database name, default: ``gcp``
+            buffer_size (int): Max number of records to hold in memeory for
+                each record_type.
+
+        """
+        # pylint: disable=too-many-instance-attributes
+
+        self._mongodb_host = host
+        self._mongodb_port = port
+        self._mongodb_username = username
+        self._mongodb_password = password
+        self._db = db
+
+        self._buffer_size = buffer_size
+        self._buffer = {}
+
+        if not (username and password):
+            self._client = MongoClient(host=self._mongodb_host,
+                                       port=self._mongodb_port,
+                                       connect=False)
+        else:
+            self._client = MongoClient(host=self._mongodb_host,
+                                       port=self._mongodb_port,
+                                       username=username,
+                                       password=password,
+                                       authSource=self._db,
+                                       authMechanism='SCRAM-SHA-256',
+                                       connect=False)
+
+    def _flush(self, record_type, records):
+        """Perform bulk insert of buffered records into MongoDB collection.
+
+        Arguments:
+          record_type (str): Collection name in MongoDB.
+          records (list): List of records of type dict to store in collection.
+
+        """
+        _log.info('Inserting %i documents into collection %s ...',
+                  len(records), record_type)
+
+        self._client[self._db][record_type].insert_many(records)
+
+        _log.info('Saved %i documents into collection %s',
+                  len(records), record_type)
+        del records[:]
+
+    def write(self, record):
+        """Write JSON records to the MongoDB collections.
+
+        This method is called once for every ``record`` read from a
+        cloud. In implementation of a store, we simply write the
+        ``record`` in JSON as document to ``gcp`` database.  The record
+        type, i.e., ``record['record_type']`` is used to determine the
+        collection name in MongoDB.
+
+        Arguments:
+            record (dict): Data to write to the MongoDB collections.
+
+        """
+        record_type = record['record_type']
+
+        if record_type not in self._buffer:
+            self._buffer[record_type] = []
+
+        # Save the record in the buffer.
+        records = self._buffer[record_type]
+        records.append(record)
+
+        # records gets more than buffer size then flush record_type records to
+        # MongoDB
+        if len(records) >= self._buffer_size:
+            self._flush(record_type, records)
+
+    def done(self):
+        """Flush remaining records from the buffer."""
+        # Flush all the residual records buffers to mongodb collections
+        for record_type, records in self._buffer.items():
```

Comment:
iterating over a changing dictionary will cause a `RuntimeException` here. Ref: [how-to-avoid-runtimeerror-dictionary-changed-size-during-iteration-error](https://stackoverflow.com/questions/11941817/how-to-avoid-runtimeerror-dictionary-changed-size-during-iteration-error).
Snippet:
```
def done(self):
        for record_type, x in self._buffer.items():
            self._flush(record_type)
```
Response:
```
$ python3 mongo_test.py
Inserted 3 out of 3 documents for record type: users
Traceback (most recent call last):
  File "mongo_test.py", line 24, in <module>
    mongo_store()
  File "mongo_test.py", line 21, in mongo_store
    store.done()
  File "/Users/p0n002h/git/py-orm/mongo/mongostore.py", line 64, in done
    for record_type, x in self._buffer.items():
RuntimeError: dictionary changed size during iteration
```

As recommended, please use `list(self._buffer.items())` in this version